controllers/upgrade_handlers_test.go

```diff
@@ -0,0 +1,194 @@
+				return &backuprestore.BackupTracker{
+					SucceededBackups: []string{"name-success "},
+				}, nil
+			},
+			wantCtlRes: doNotRequeue(),
+			wantErr:    assert.NoError,
+		},
+		{
+			name: "backup failed",
+			wantFuncReturn: func() (*backuprestore.BackupTracker, error) {
+				return &backuprestore.BackupTracker{
+					FailedBackups: []string{"name-failed"},
+				}, nil
+			},
+			wantCtlRes: doNotRequeue(),
+			wantErr:    assert.Error,
+		},
+		{
+			name: "backup pending",
+			wantFuncReturn: func() (*backuprestore.BackupTracker, error) {
+				return &backuprestore.BackupTracker{
+					PendingBackups: []string{"name-pending"},
+				}, nil
+			},
+			wantCtlRes: requeueWithMediumInterval(),
+			wantErr:    assert.NoError,
+		},
+		{
+			name: "backup progressing",
+			wantFuncReturn: func() (*backuprestore.BackupTracker, error) {
+				return &backuprestore.BackupTracker{
+					ProgressingBackups: []string{"name-progressing"},
+				}, nil
+			},
+			wantCtlRes: requeueWithCustomInterval(5 * time.Second),
+			wantErr:    assert.NoError,
+		},
+		{
+			name: "backup failed validation by oadp",
+			wantFuncReturn: func() (*backuprestore.BackupTracker, error) {
+				return &backuprestore.BackupTracker{
+					FailedValidationBackups: []string{"name-failed-validation"},
+				}, nil
+			},
+			wantCtlRes: requeueWithMediumInterval(),
+			wantErr:    assert.NoError,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			//setup
+			velS := [][]*velerov1.Backup{{&velerov1.Backup{}}}
+			mockBackuprestore.EXPECT().GetSortedBackupsFromConfigmap(gomock.Any(), gomock.Any()).Return(velS, nil).AnyTimes()
+
+			r := &ImageBasedUpgradeReconciler{
+				Log:           logr.Logger{},
+				BackupRestore: mockBackuprestore,
+			}
+
+			mockBackuprestore.EXPECT().StartOrTrackBackup(context.Background(), velS[0]).Return(tt.wantFuncReturn()).Times(1)
+
+			// assert
+			got, err := r.handleBackup(context.Background(), &lcav1alpha1.ImageBasedUpgrade{})
+			if !tt.wantErr(t, err, fmt.Sprintf("handleBackup")) {
+				return
+			}
+			assert.Equalf(t, tt.wantCtlRes.RequeueAfter, got.RequeueAfter, "ctl interval: handleBackup")
+
+		})
+	}
+}
+
+func TestImageBasedUpgradeReconciler_handleRestore(t *testing.T) {
+	mockController := gomock.NewController(t)
+	mockBackuprestore := mock_backuprestore.NewMockBackuperRestorer(mockController)
+	defer func() {
+		defer mockController.Finish()
+	}()
+
+	tests := []struct {
+		name           string
+		wantFuncReturn func() (*backuprestore.RestoreTracker, error)
+		wantCtlRes     controllerruntime.Result
+		wantErr        assert.ErrorAssertionFunc
+		ibuStatus      lcav1alpha1.ImageBasedUpgradeStatus
+	}{
+		{
+			name: "restore successful",
+			wantFuncReturn: func() (*backuprestore.RestoreTracker, error) {
+				return &backuprestore.RestoreTracker{
+					SucceededRestores: []string{"name-success"},
+				}, nil
+			},
+			wantCtlRes: doNotRequeue(),
```

Comment:
Probably consider adding testcases for two restore groups?
Two groups of restores,  the first succeeded + the second succeeded = succeeded = doNotRequeue()
                                           the first succeeded + the second failed = failed = doNotRequeue()

Reply:
Added!